libcudacxx/include/cuda/std/__numeric/saturation_arithmetic.h

```diff
@@ -0,0 +1,802 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX___NUMERIC_SATURATION_ARITHMETIC_H
+#define _LIBCUDACXX___NUMERIC_SATURATION_ARITHMETIC_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if _CCCL_STD_VER >= 2017
+
+#  include <cuda/std/__algorithm/clamp.h>
+#  include <cuda/std/__algorithm/max.h>
+#  include <cuda/std/__algorithm/min.h>
+#  include <cuda/std/__concepts/concept_macros.h>
+#  include <cuda/std/__type_traits/is_arithmetic_integral.h>
+#  include <cuda/std/__type_traits/is_constant_evaluated.h>
+#  include <cuda/std/__type_traits/is_signed.h>
+#  include <cuda/std/__type_traits/is_unsigned.h>
+#  include <cuda/std/__type_traits/make_unsigned.h>
+#  include <cuda/std/__type_traits/remove_cv.h>
+#  include <cuda/std/__utility/cmp.h>
+#  include <cuda/std/climits>
+#  include <cuda/std/cstdint>
+#  include <cuda/std/limits>
+
+#  include <nv/target>
+
+#  if _CCCL_COMPILER(MSVC)
+#    include <intrin.h>
+#  endif // _CCCL_COMPILER(MSVC)
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+class __add_sat
+{
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp
+  __clamp_overflow(_Tp, _Tp __y, _Tp __result, bool __overflow) noexcept
+  {
+    if (__overflow)
+    {
+      if constexpr (_CCCL_TRAIT(is_unsigned, _Tp))
+      {
+        __result = _CUDA_VSTD::numeric_limits<_Tp>::max();
+      }
+      else
+      {
+        __result = (__y < _Tp{0}) ? _CUDA_VSTD::numeric_limits<_Tp>::min() : _CUDA_VSTD::numeric_limits<_Tp>::max();
+      }
+    }
+    return __result;
+  }
+
+public:
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __impl_generic(_Tp __x, _Tp __y) noexcept
+  {
+    if constexpr (_CCCL_TRAIT(is_signed, _Tp))
+    {
+      using _Up    = make_unsigned_t<_Tp>;
+      _Tp __result = static_cast<_Tp>(static_cast<_Up>(__x) + static_cast<_Up>(__y));
+      return __clamp_overflow(__x, __y, __result, (__result < __x) == !(__y < static_cast<_Tp>(0)));
+    }
+    else
+    {
+      _Tp __result = static_cast<_Tp>(__x + __y);
+      return __clamp_overflow(__x, __y, __result, (__result < __x));
+    }
+  }
+
+#  if defined(_CCCL_BUILTIN_ADD_OVERFLOW)
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __impl_builtin(_Tp __x, _Tp __y) noexcept
+  {
+    _Tp __result{};
+    bool __overflow = _CCCL_BUILTIN_ADD_OVERFLOW(__x, __y, &__result);
+
+    return __clamp_overflow(__x, __y, __result, __overflow);
+  }
+#  endif // _CCCL_BUILTIN_ADD_OVERFLOW
+
+  template <class _Tp>
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_HOST static _Tp __impl_host(_Tp __x, _Tp __y) noexcept
```

Comment:
We need to guard this for nvrtc because it naturally cannot handle host functions